### PR TITLE
feat: block/mute user

### PR DESCRIPTION
**Pull request by swkatmask**
MF-0000 FW-0000

### CODE REVIEW

File: src/store/usePostStore.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    persist(
+        immer((set) => ({
+            allBlockedUsers: {},
+            blockUser: (operator: Profile, profile: Profile) => {
+                set((state) => {
+                    const key = `${operator.source}:${operator.profileId}`;
+                    if (!state.allBlockedUsers[key]) {
+                        state.allBlockedUsers[key] = [];
+                    }
+                    state.allBlockedUsers[key].push(profile.profileId);
```
Comment (guanbinrui) [inline src/store/usePostStore.ts:23]: Dedup

File: src/hooks/useBlockUser.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * Block a user
+ */
+export function useBlockUser(operator: Profile | null) {
+    const { blockUser } = usePostState();
+    return useAsyncFn(
+        async (profile: Profile) => {
+            try {
+                const provider = resolveSocialMediaProvider(profile.source);
+                const result = await provider.blockUser(profile.profileId);
+                if (result && operator) blockUser(operator, profile);
```
Comment (guanbinrui) [inline src/hooks/useBlockUser.ts:19]: Did you miss it?

```suggestion
                if (result && operator) await blockUser(operator, profile);
```

File: src/providers/warpcast/SocialMedia.ts [hunk 3]
```diff
@@ -642,11 +642,16 @@ class WarpcastSocialMedia implements Provider {
         );
     }
     async reportUser(profileId: string): Promise<boolean> {
-        throw new Error('Method not implemented.');
+        // TODO Mocking result for now.
+        return true;
```
Comment (guanbinrui) [inline src/providers/warpcast/SocialMedia.ts:646]: Unreachable? So keep it throws not implement erorr.

File: src/hooks/useReportUser.ts [hunk 4]
```diff
@@ -4,19 +4,25 @@ import { useAsyncFn } from 'react-use';
...
-        }
-    }, []);
+export function useReportUser(operator: Profile | null) {
+    const { blockUser } = usePostState();
+    return useAsyncFn(
+        async (profile: Profile) => {
+            try {
+                const provider = resolveSocialMediaProvider(profile.source);
+                const result = await provider.reportUser(profile.profileId);
+                if (result && operator) blockUser(operator, profile);
```
Comment (guanbinrui) [inline src/hooks/useReportUser.ts:19]:
```suggestion
                if (result && operator) reportUser(operator, profile);
```

File: src/hooks/useBlockUser.ts (hunk 2 above)
Comment (swkatmask) [reply src/hooks/useBlockUser.ts:19]: No, this is blocking inside state store.

File: src/hooks/useReportUser.ts (hunk 4 above)
Comment (swkatmask) [reply src/hooks/useReportUser.ts:19]: Block in state store.

File: src/store/usePostStore.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                    if (!state.allBlockedUsers[key]) {
+                        state.allBlockedUsers[key] = [];
+                    }
+                    const prevList = get().allBlockedUsers[key] || [];
+                    state.allBlockedUsers[key] = uniq([...prevList, profile.profileId]);
+                });
+            },
+        })),
+        {
+            name: 'firefly-posts',
```
Comment (guanbinrui) [inline src/store/usePostStore.ts:30]:
```suggestion
            name: 'blocked-users',
```

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**